src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -439,22 +619,45 @@ double LMNNFunction<MetricType>::EvaluateWithGradient(
         // Calculate cost due to {data point, target neighbors, impostors}
         // triplets.
         double eval = 0;
-        if ((iteration - 1 % range == 0) && (iteration - 1 != 0))
+
+        // Flag to trigger exact eval calculation.
+        bool exactEval = true;
+
+        // Bounds for eval.
+        if (transformationOld.n_elem != 0 && !std::isnan(evalOld[i](j, l)))
         {
-          eval = metric.Evaluate(transformedDataset.col(i),
-                    transformedDataset.col(targetNeighbors(j, i))) -
-                 distance(l, i);
+          eval = evalOld[i](j, l) + transformationDiff *
+            (norm(targetNeighbors(j, i)) + norm(impostors(l, i)) +
+            2 * norm(i));
+
+          // Check if there is need to calculate exact eval value.
+          if (eval <= -1)
+            exactEval = false;
         }
-        else
+
+        // Calculate exact eval value.
+        if(exactEval)
         {
-          eval = metric.Evaluate(transformedDataset.col(i),
-                    transformedDataset.col(targetNeighbors(j, i))) -
-                 metric.Evaluate(transformedDataset.col(i),
-                    transformedDataset.col(impostors(l, i)));
+          if (iteration - 1 % range == 0)
+          {
+            eval = metric.Evaluate(transformedDataset.col(i),
+                     transformedDataset.col(targetNeighbors(j, i))) -
+                 distance(l, i);
+          }
+          else
+          {
+            eval = metric.Evaluate(transformedDataset.col(i),
+                     transformedDataset.col(targetNeighbors(j, i))) -
+                   metric.Evaluate(transformedDataset.col(i),
+                       transformedDataset.col(impostors(l, i)));
+          }
         }
 
+        // Update cache eval value.
+        evalOld[i](j, l) = eval;
```

Comment:
Is it actually a problem if we store an approximate bounded value in evalOld?

Reply:
I think it can be when approximate eval is very close to -1 than actual eval or vice-versa.

Reply:
At the very least we won't be getting same results.

Reply:
But does your idea here match the results I was showing in my simulation?  Note that I was comparing the output transformation matrices and received the exact same results.  To put it more clearly: suppose that we take `evalOld[i](j, l) = eval + x` for some x greater than 0; now, the bound for that particular `evalOld` is loose.  What will happen the next iteration?

Reply:
Sorry, that was just a thought because of which earlier I was updating it after break statement. If the results don't divert then I think we shouldn't worry about this. Truly, I also haven't seen any deviation but was fearing it might happen with some different dataset.

Reply:
Don't worry, it's definitely ok and a good thing to have the thought, but it's worthwhile to follow through on the thinking and make sure we're on the same page.  Do you mind working out what happens the next iteration if `evalOld` is loose like I wrote above?

Reply:
I think I haven't got that part properly. Just to be sure: loose like - we don't calculate exact value if we have calculated approximate one?

Reply:
Or do you mean just any random x > 0 to be placed there?

Reply:
Right, it turns out these are the same things, but I would say it's easier to think about if you just say that we've taken `evalOld[i](j, l) = eval + x` for some `x` greater than 0.  Later we can generalize to what happens if we've taken some approximate value.

Reply:
Without x:

Iris:
```
Evaluate(): 0 pruned of 2383, transformation diff 0.  Active: 1949, close 420, inactive 14.
Evaluate(): 0 pruned of 1782, transformation diff 0.998596.  Active: 1245, close 251, inactive 286.
Evaluate(): 0 pruned of 1655, transformation diff 0.492509.  Active: 1074, close 258, inactive 323.
Evaluate(): 0 pruned of 1339, transformation diff 1.34492.  Active: 685, close 244, inactive 410.
Evaluate(): 0 pruned of 1475, transformation diff 0.672462.  Active: 847, close 249, inactive 379.
Evaluate(): 0 pruned of 1364, transformation diff 0.47076.  Active: 710, close 243, inactive 411.
Evaluate(): 0 pruned of 1411, transformation diff 0.23538.  Active: 771, close 244, inactive 396.
Evaluate(): 233 pruned of 1450, transformation diff 0.11769.  Active: 818, close 324, inactive 308.
Evaluate(): 78 pruned of 1460, transformation diff 0.058845.  Active: 829, close 312, inactive 319.
Evaluate(): 321 pruned of 1467, transformation diff 0.0294225.  Active: 838, close 355, inactive 274.
Evaluate(): 370 pruned of 1472, transformation diff 0.0147113.  Active: 844, close 294, inactive 334.
Evaluate(): 443 pruned of 1473, transformation diff 0.00735563.  Active: 845, close 304, inactive 324.
Evaluate(): 485 pruned of 1473, transformation diff 0.00367782.  Active: 845, close 319, inactive 309.
Evaluate(): 532 pruned of 1473, transformation diff 0.00183891.  Active: 845, close 317, inactive 311.
Evaluate(): 560 pruned of 1475, transformation diff 0.000919454.  Active: 847, close 321, inactive 307.
Evaluate(): 579 pruned of 1475, transformation diff 0.000459727.  Active: 847, close 321, inactive 307.
Evaluate(): 607 pruned of 1475, transformation diff 0.000229863.  Active: 847, close 321, inactive 307.
Evaluate(): 612 pruned of 1475, transformation diff 0.000114932.  Active: 847, close 322, inactive 306.
Evaluate(): 622 pruned of 1475, transformation diff 5.74659e-05.  Active: 847, close 322, inactive 306.
Evaluate(): 623 pruned of 1475, transformation diff 2.87329e-05.  Active: 847, close 322, inactive 306.
Evaluate(): 626 pruned of 1475, transformation diff 1.43665e-05.  Active: 847, close 322, inactive 306.
Evaluate(): 627 pruned of 1475, transformation diff 7.18323e-06.  Active: 847, close 322, inactive 306.
Evaluate(): 626 pruned of 1475, transformation diff 3.59162e-06.  Active: 847, close 322, inactive 306.
Evaluate(): 628 pruned of 1475, transformation diff 1.79581e-06.  Active: 847, close 322, inactive 306.
Evaluate(): 628 pruned of 1475, transformation diff 8.97904e-07.  Active: 847, close 322, inactive 306.
Evaluate(): 628 pruned of 1475, transformation diff 4.48952e-07.  Active: 847, close 322, inactive 306.
```
vc2:
```
Evaluate(): 0 pruned of 4223, transformation diff 0.0583577.  Active: 3931, close 157, inactive 135.
Evaluate(): 0 pruned of 4417, transformation diff 0.0457136.  Active: 4187, close 207, inactive 23.
Evaluate(): 0 pruned of 4457, transformation diff 0.041658.  Active: 4212, close 187, inactive 58.
Evaluate(): 0 pruned of 4450, transformation diff 0.0165774.  Active: 4224, close 192, inactive 34.
Evaluate(): 0 pruned of 4484, transformation diff 0.00613733.  Active: 4265, close 190, inactive 29.
Evaluate(): 0 pruned of 4520, transformation diff 0.00481902.  Active: 4316, close 178, inactive 26.
Evaluate(): 0 pruned of 4501, transformation diff 0.00240951.  Active: 4284, close 191, inactive 26.
Evaluate(): 27 pruned of 4491, transformation diff 0.00120476.  Active: 4273, close 217, inactive 1.
Evaluate(): 52 pruned of 4487, transformation diff 0.000602378.  Active: 4267, close 199, inactive 21.
Evaluate(): 92 pruned of 4484, transformation diff 0.000301189.  Active: 4265, close 206, inactive 13.
Evaluate(): 118 pruned of 4483, transformation diff 0.000150594.  Active: 4264, close 206, inactive 13.
Evaluate(): 143 pruned of 4484, transformation diff 7.52972e-05.  Active: 4265, close 208, inactive 11.
Evaluate(): 181 pruned of 4484, transformation diff 3.76486e-05.  Active: 4265, close 209, inactive 10.
Evaluate(): 194 pruned of 4484, transformation diff 1.88243e-05.  Active: 4265, close 210, inactive 9.
Evaluate(): 202 pruned of 4484, transformation diff 9.41215e-06.  Active: 4265, close 210, inactive 9.
Evaluate(): 211 pruned of 4484, transformation diff 4.70608e-06.  Active: 4265, close 210, inactive 9.
Evaluate(): 212 pruned of 4484, transformation diff 2.35304e-06.  Active: 4265, close 210, inactive 9.
Evaluate(): 213 pruned of 4484, transformation diff 1.17652e-06.  Active: 4265, close 210, inactive 9.
Evaluate(): 217 pruned of 4484, transformation diff 5.8826e-07.  Active: 4265, close 210, inactive 9.
Evaluate(): 219 pruned of 4484, transformation diff 2.9413e-07.  Active: 4265, close 210, inactive 9.
Evaluate(): 219 pruned of 4484, transformation diff 1.47065e-07.  Active: 4265, close 210, inactive 9.
Evaluate(): 219 pruned of 4484, transformation diff 7.35324e-08.  Active: 4265, close 210, inactive 9.
Evaluate(): 219 pruned of 4484, transformation diff 3.67662e-08.  Active: 4265, close 210, inactive 9.
Evaluate(): 219 pruned of 4484, transformation diff 1.83831e-08.  Active: 4265, close 210, inactive 9.
Evaluate(): 219 pruned of 4484, transformation diff 9.19156e-09.  Active: 4265, close 210, inactive 9.
Evaluate(): 219 pruned of 4484, transformation diff 4.59578e-09.  Active: 4265, close 210, inactive 9.
```

with a random x:
Iris:
```
Evaluate(): 0 pruned of 3008, transformation diff 0.  Active: 2697, close 311, inactive 0.
Evaluate(): 0 pruned of 1956, transformation diff 0.998204.  Active: 1457, close 234, inactive 265.
Evaluate(): 0 pruned of 1877, transformation diff 0.231619.  Active: 1364, close 230, inactive 283.
Evaluate(): 0 pruned of 1795, transformation diff 0.254781.  Active: 1259, close 238, inactive 298.
Evaluate(): 0 pruned of 1135, transformation diff 5.42563.  Active: 436, close 184, inactive 515.
Evaluate(): 0 pruned of 1271, transformation diff 2.71282.  Active: 588, close 226, inactive 457.
Evaluate(): 0 pruned of 1276, transformation diff 0.633136.  Active: 597, close 227, inactive 452.
Evaluate(): 0 pruned of 1348, transformation diff 0.52932.  Active: 688, close 239, inactive 421.
Evaluate(): 63 pruned of 1365, transformation diff 0.178184.  Active: 711, close 305, inactive 349.
Evaluate(): 106 pruned of 1347, transformation diff 0.148173.  Active: 691, close 345, inactive 311.
Evaluate(): 153 pruned of 1356, transformation diff 0.0740863.  Active: 701, close 273, inactive 382.
Evaluate(): 277 pruned of 1361, transformation diff 0.0370431.  Active: 707, close 368, inactive 286.
Evaluate(): 271 pruned of 1363, transformation diff 0.0185216.  Active: 708, close 354, inactive 301.
Evaluate(): 336 pruned of 1364, transformation diff 0.00926078.  Active: 709, close 342, inactive 313.
Evaluate(): 393 pruned of 1364, transformation diff 0.00463039.  Active: 709, close 399, inactive 256.
Evaluate(): 371 pruned of 1364, transformation diff 0.0023152.  Active: 709, close 398, inactive 257.
Evaluate(): 380 pruned of 1365, transformation diff 0.0011576.  Active: 710, close 381, inactive 274.
Evaluate(): 401 pruned of 1365, transformation diff 0.000578799.  Active: 711, close 363, inactive 291.
Evaluate(): 408 pruned of 1365, transformation diff 0.000289399.  Active: 711, close 377, inactive 277.
Evaluate(): 398 pruned of 1365, transformation diff 0.0001447.  Active: 711, close 380, inactive 274.
```

vc2:
```
Evaluate(): 0 pruned of 4535, transformation diff 0.0519045.  Active: 4351, close 147, inactive 37.
Evaluate(): 0 pruned of 4470, transformation diff 0.00897.  Active: 4259, close 165, inactive 46.
Evaluate(): 0 pruned of 4485, transformation diff 0.00883794.  Active: 4276, close 175, inactive 34.
Evaluate(): 0 pruned of 4486, transformation diff 0.00396619.  Active: 4272, close 180, inactive 34.
Evaluate(): 0 pruned of 4505, transformation diff 0.003486.  Active: 4295, close 180, inactive 30.
Evaluate(): 6 pruned of 4497, transformation diff 0.001743.  Active: 4285, close 185, inactive 27.
Evaluate(): 25 pruned of 4490, transformation diff 0.000871501.  Active: 4276, close 196, inactive 18.
Evaluate(): 39 pruned of 4488, transformation diff 0.00043575.  Active: 4274, close 194, inactive 20.
Evaluate(): 43 pruned of 4486, transformation diff 0.000217875.  Active: 4272, close 199, inactive 15.
Evaluate(): 60 pruned of 4486, transformation diff 0.000108938.  Active: 4272, close 199, inactive 15.
Evaluate(): 64 pruned of 4486, transformation diff 5.44688e-05.  Active: 4272, close 199, inactive 15.
Evaluate(): 64 pruned of 4486, transformation diff 2.72344e-05.  Active: 4272, close 199, inactive 15.
Evaluate(): 54 pruned of 4486, transformation diff 1.36172e-05.  Active: 4272, close 190, inactive 24.
Evaluate(): 62 pruned of 4486, transformation diff 6.8086e-06.  Active: 4272, close 199, inactive 15.
Evaluate(): 70 pruned of 4486, transformation diff 3.4043e-06.  Active: 4272, close 203, inactive 11.
Evaluate(): 60 pruned of 4486, transformation diff 1.70215e-06.  Active: 4272, close 200, inactive 14.
Evaluate(): 54 pruned of 4486, transformation diff 8.51075e-07.  Active: 4272, close 192, inactive 22.
Evaluate(): 67 pruned of 4486, transformation diff 4.25537e-07.  Active: 4272, close 201, inactive 13.
Evaluate(): 77 pruned of 4486, transformation diff 2.12769e-07.  Active: 4272, close 197, inactive 17.
Evaluate(): 61 pruned of 4486, transformation diff 1.06384e-07.  Active: 4272, close 195, inactive 19.
Evaluate(): 65 pruned of 4486, transformation diff 5.31922e-08.  Active: 4272, close 194, inactive 20.
Evaluate(): 77 pruned of 4486, transformation diff 2.65961e-08.  Active: 4272, close 199, inactive 15.
Evaluate(): 66 pruned of 4486, transformation diff 1.3298e-08.  Active: 4272, close 201, inactive 13.

```
I don't think there is much deviation here.

Reply:
So I can see from the results that fewer points are pruned, but what I was hoping to have you think about is theoretically what should happen when we add some `x` to the `eval` value.  Will the bound still be valid the next iteration?  Consider taking a look at the derivation of this bound and plugging in the change of adding `x` to it.

Reply:
Here's the theoretical essence as per my thought -
We have `eval < evalOld + pointSpecificConstantTerm` So, if we are increasing evalOld by adding a positive x, we are basically increasing upper bound for new eval. It will be `eval + x`. Hence the calculation for eval bound which we are doing will give `eval` increased by a factor of `x`, hence pulling it towards positive axis or just reducing its probability of being negative. Which all together means less number of prunes.
Please tell me if it sounds a bit confusing. I can try to clarify further.

Reply:
The explanation makes sense to me.  I agree---increasing `x` would result in fewer prunes.  Would you then conclude that for any positive `x`, even if we take `evalOld[i](j, l) = eval + x`, the bound will still be correct?  (As in, it will still be a valid bound.)

Reply:
No doubt, It will surely be correct. Just the prunes will decrease.

Reply:
Having loose bounds definitely shouldn't be wrong.

Reply:
Ok, I agree.  Now we can get back to your original question about the placement of how `evalOld` is assigned.  Even if we assign `evalOld` before the prune, the worst case is that we assign a bound instead of an inexact value (so, this is equivalent to saying `evalOld[i](j, l) = eval + x` where `x` is the bounding term || L_{t + 1} - L_t || * (|| x_j || + || x_l || + 2 || x_i ||).  Since we both agree that it is still a valid bound, just possibly loose, there is no problem with doing this.  In fact, in the next iteration, if the exact calculation of `eval` is not done and we end up pruning based on another bound, our stored `evalOld` value will be loose---but not invalid.  So we will still get some prunes, but it will not be perfect.

The other option, as you pointed out, is to assign `evalOld` after the pruning.  But in this case, this means that we will never set any `evalOld` value for a lot of pruned points, and then it won't be possible to prune them in the following iteration.  You can see this in the simulations I ran yesterday---without setting `evalOld` before the pruning, we get zero prunes at all throughout the entire algorithm (and no speedup).

So, given these two choices, it's best to go with the approach that will give some prunes despite the bound being looser than it could possibly be.

Reply:
Totally Agree :) Thanks for clearing this out.